btf/btf.go

```diff
@@ -695,5 +695,11 @@ func (iter *TypesIterator) Next() bool {
 	iter.Type, ok = iter.spec.typeByID(iter.id)
 	iter.id++
 	iter.done = !ok
+	if !iter.done {
+		// Skip declTags
+		if _, ok := iter.Type.(*declTag); ok {
+			return iter.Next()
+		}
+	}
```

Comment:
Should we modify `indexTypes` as well?

How about `fixup` in `readAndInflateTypes`, should we allow (invalid) references to `DeclTag`s in contexts where proper TypeID is expected (e.g. a struct member referencing `DeclTag` instead of a type)?

It is not clear to me whether this filtering is done for user convenience OR to avoid dealing with pre-existing `DeclTag`s during marshalling.

Reply:
The situation as it stands is that the iterator returns `*declTag` since its is a `Type` though you can't do anything with it outside of printing it. Because this PR causes decl tags to be merged into the types they refer to, we essentially make holes in the type list:
```
BTF blob
0: void
1: int
2: declTag -> int
3: struct

Go:
0: void
1: int + declTag
3: struct
```

We pretty much represent a BTF blob as `[]Type` everywhere, where the ID of the type is the index into the slice, which in turn match the ID in lets say BTF.ext. So when we fold declTags into other types we need to still fill the `[]Type` slice with something to not mess with the indexes/IDs. I discovered this when initially trying to just not emit a `Type` for declTags. But its causes problem, for anything outside BTF to reference IDs such as relocations or func info.

I tried just making `nil` entries, but we assume every `Type` in the slice is always non-nil, so I was afraid that this would introduce unseen bugs and nil-dereferences. 

So, I came up with 3 ways to work around it:

1. Keep emitting `*declTag` , but just hide them during iteration. That is the smallest change. Users can technically still access them via the TypeByID functions ect, but there is nothing that references declTag IDs so that doesn't seem like a big deal
2. Introduce a new `Type` implementation like "hiddenType" or something, so its more explicit. But given we have declTag already, It seemed excessive
3. Emit a `Void` for every *declTag. Normally only at index 0, it basically means "nothing". However, without similar modifications to the iterator you get a whole bunch of `Void`s which seems really odd and unexpected.

I went for number 1, but if you see a better way of dealing with this then I am open to suggestions.

Reply:
Option 1 seems like the best compromise to me. We're doing our best to abstract over any warts and make things as user-friendly as possible, but if an object has a type ID, it would be surprising if it didn't in our implementation. We can try to limit what the user can do with a `declTag`, but hiding it completely by relegating it to the BTF encoding would require significant internal (and external?) changes.

Essentially, this would mean we need to tolerate holes in a type ID sequence, which goes against many assumptions made by the code. Let's go with this for now. Callers can still pull a `declTag` out of a Spec if they really want to, e.g. for debugging purposes, but it won't be of much use. This was the case before as well.

Let's see if any other declTag-like things get merged down the line, maybe there will be more of these internal types that need special treatment. We can reevaluate if/when that time comes.